Add support for Oreo Autofill

FabianHenneke opened the pull request:
## :loudspeaker: Type of change
<!--- Put an `x` in the boxes that apply -->
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring


## :scroll: Description
<!--- Describe your changes in detail -->
This PR adds an Oreo Autofill service to Password Store that offers to fill forms and save credentials (username and password) in compatible apps and browsers. It allows users to

- fill from Password Store entries, selected either via free search or via a matching system;
- generate new credentials and save them in their repository;
- save credentials they manually entered in login forms.

The Autofill implementation aims to mitigate vulnerabilities present in many other password managers by strictly adhering to the following guidelines with regards to how it matches store entries to apps and websites:

- If a trusted browser app (identified via package name and certificate hash) annotates a login form with a web domain, the eTLD + 1 (also called "canonical domain") is associated with all entries the user chooses to fill, generate or save.
- If a login form does not originate from a website displayed in a trusted browser app, it is associated with the app itself (identified via package name and certificate hash). If, at any later point in time, the certificate hash changes, the user is warned about a potential phishing attempt.

Browsers vary in their level of support for Autofill, which generally falls into one of the following four categories:

1. **No support**: For whatever reason, some browsers (e.g. Opera) never trigger any Autofill requests.
2. **Single-origin support**: Many browsers such as Chrome and its derivatives support Autofill in compatibility mode, but fail to report all web origins present on a page (such as in iframes). Since filling into non-visible form elements across origin boundaries is ripe for abuse, in such browsers Autofill is limited to filling a single password into the focused password field (plus a username). I hope that this strikes a decent balance between usability and security.
3. **Multi-origin support**: Browsers with custom Autofill support or those that rely on WebViews correctly propagate origin information to all form elements, which allows the service to fill more aggressively - for example, it can offer to fill a registration form with a username and two new-password fields in a single action, regardless of focus. Mozilla browsers, Opera Mini/Touch, DuckDuckGo and the Tor Browser fall in this category.
4. **Save support**: A few browsers have already added custom Autofill support and thus also provide submitted credentials for us to save. This e.g. works in Firefox Focus, Opera Mini/Touch and DuckDuckGo.

Of course, in order to be really helpful, Password Store needs to detect login forms as precisely as possible. For this reason, I have implemented a [Kotlin DSL for Autofill rules](https://github.com/FabianHenneke/Android-Password-Store/blob/fhenneke_autofill_pr/app/src/main/java/com/zeapo/pwdstore/autofill/oreo/AutofillStrategy.kt) that aims to be accessible to a broad range of developers. Please let me know about your thoughts on this in particular.

## :bulb: Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This fixes #410 and may supersede #551.

I took #551 by @c4software as the starting point and would like to thank them very much for getting the ball rolling. 

## :green_heart: How did you test it?
I tested the service with many apps and websites. Since this is a big feature that can potentially interact with any app or website out there, this does of course require much more testing than I cannot take on myself.

## :pencil: Checklist
<!--- Put an `x` in the boxes that apply -->
- [x] I ran `./gradlew spotlessApply` before submitting the PR
- [x] I reviewed submitted code

## :crystal_ball: Next steps
- [x] Decide on a naming convention to prevent confusion with Accessibility autofill
- [ ] Add UI that introduces the user to the new Autofill API and allows to enable it from within the app
- [x] (Optional) Add Autofill-specific settings to the Password Store settings screen
- [ ] Test with various apps and websites and add Autofill rules to AutofillStrategy.kt
- [ ] Verify the authenticity of the browser certificate hashes and the browser compatibility settings
- [ ] Add translations (I can cover German if needed)

## :camera_flash: Screenshots / GIFs
<!--- Mandatory for UI changes -->
<!-- Uncomment the next line and replace it with links to your screenshots. -->

![Screenshot_20200318-133737](https://user-images.githubusercontent.com/4312191/76979738-a7a64c00-6938-11ea-8510-75a079f86865.jpg)
![Screenshot_20200318-133747](https://user-images.githubusercontent.com/4312191/76979737-a70db580-6938-11ea-833e-849e6c716470.jpg)
![Screenshot_20200318-144805](https://user-images.githubusercontent.com/4312191/76979733-a6751f00-6938-11ea-88a6-80f1f7acbcd8.jpg)
![Screenshot_20200318-144842](https://user-images.githubusercontent.com/4312191/76979731-a5dc8880-6938-11ea-9e0c-8b9c059683e3.jpg)
![Screenshot_20200318-164602](https://user-images.githubusercontent.com/4312191/76979711-9f4e1100-6938-11ea-8ce9-fc2d2970b695.jpg)
![Screenshot_20200318-153116](https://user-images.githubusercontent.com/4312191/76979725-a4ab5b80-6938-11ea-88aa-cf42805a96ee.jpg)
![Screenshot_20200318-153122](https://user-images.githubusercontent.com/4312191/76979720-a37a2e80-6938-11ea-86e3-ddcfe8ba5a40.jpg)

c4software commented:
Awesome 👍👍👍

msfjarvis commented:
This is amazing work, and I deeply appreciate the amount of thought that clearly has gone behind this. Our next release is scheduled for the 21st of March, which does not leave enough time to both review and test this so I've set the timeline for this for April instead. I'll get back in an hour or so and do a thorough review as well as provide the feedback you asked for. Thanks again for taking the time to contribute to APS, we're extremely grateful :)

FabianHenneke commented:
> This is amazing work, and I deeply appreciate the amount of thought that clearly has gone behind this. Our next release is scheduled for the 21st of March, which does not leave enough time to both review and test this so I've set the timeline for this for April instead. I'll get back in an hour or so and do a thorough review as well as provide the feedback you asked for. Thanks again for taking the time to contribute to APS, we're extremely grateful :)

Thanks, this sounds like a good plan! I am using Password Store for both private and work accounts, multiple times a day, so the effort is definitely worth it for me and hopefully also for others.

Note that I just force-pushed a change to satisfy the Lint CI checks.

Skrilltrax commented:
Thanks for the PR, this is a huge feature for so many users. We Really appreciate your work on this 😁

msfjarvis commented on app/src/main/java/com/zeapo/pwdstore/autofill/oreo/ui/AutofillFilterActivity.kt:
```diff
@@ -0,0 +1,199 @@
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        super.onActivityResult(requestCode, resultCode, data)
+        if (requestCode == 1) {
```
Probably worth moving into a static `const val`

msfjarvis commented on app/src/main/java/com/zeapo/pwdstore/autofill/oreo/ui/AutofillFilterActivity.kt:
```diff
@@ -0,0 +1,199 @@
+        recursiveFilter(initialFilter, strict = formOrigin is FormOrigin.Web)
+
+        shouldMatch.apply {
+            text = getString(
```
ditto here

msfjarvis commented on app/src/main/java/com/zeapo/pwdstore/autofill/oreo/ui/AutofillFilterActivity.kt:
```diff
@@ -0,0 +1,199 @@
+                onClick { decryptAndFill(item) }
+            }
+        }
+        rvPassword.apply {
```
I personally don't see the point of using `.apply` when we're only changing one property

msfjarvis commented on app/src/main/java/com/zeapo/pwdstore/autofill/oreo/ui/AutofillSaveActivity.kt:
```diff
@@ -0,0 +1,140 @@
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        val repo = PasswordRepository.getRepositoryDirectory(applicationContext)
+        val username: String? = intent.getStringExtra(EXTRA_USERNAME)
```
I believe this automatically gets typed to `String?`

msfjarvis commented on app/src/main/java/com/zeapo/pwdstore/crypto/PgpActivity.kt:
```diff
@@ -81,6 +82,10 @@ class PgpActivity : AppCompatActivity(), OpenPgpServiceConnection.OnBound {
 
+    private val suggestedPass: String? by lazy { intent.getStringExtra("SUGGESTED_PASS") }
+    private val suggestedExtra: String? by lazy { intent.getStringExtra("SUGGESTED_EXTRA") }
+    private val shouldGeneratePassword: Boolean by lazy { intent.getBooleanExtra("GENERATE_PASSWORD", false) }
```
All these fields should also get typed correctly from the framework

msfjarvis commented:
Unfortunately I haven't had much luck in testing this out. I installed a release build, set Autofill Service in Android Settings to Password Store and am yet to see the UI. I've tried it on spotify.com, paypal.com, as well as fabianhenneke.github.io/Android-Password-Store without success.

FabianHenneke commented:
Thanks for the comments, I will take a look!

May I ask which browser you are using? Chrome's Autofill support is quite flaky and sometimes just stops working until the browser is killed and restarted.

msfjarvis commented:
Indeed I was using Chrome and just like you said, autofill is flaky. Started working literally half a minute ago. Added a mental note to also notify users of this when we build out the onboarding UI.

FabianHenneke commented:
> Indeed I was using Chrome and just like you said, autofill is flaky. Started working literally half a minute ago. Added a mental note to also notify users of this when we build out the onboarding UI.

You can refer to https://bugs.chromium.org/p/chromium/issues/detail?id=1015381. 

It's unfortunate that Google's own browser is currently the least capable in terms of Autofill, but at least there is reasonable hope that they will natively support Autofill in the future.

FabianHenneke commented on app/build.gradle:
```diff
@@ -96,13 +96,16 @@ dependencies {
     implementation deps.third_party.commons_io
     implementation deps.third_party.commons_codec
     implementation deps.third_party.fastscroll
+    implementation deps.third_party.guava
```
I agree that this may not be worth it, it's just what the Autofill sample code recommended. I did some research and would now want to switch to one of [Mozilla's Android Components](https://github.com/mozilla-mobile/android-components/blob/master/components/lib/publicsuffixlist/README.md). What do you think?

msfjarvis replied:
Looks like it does all we need it to, and does that in a slim package. Please give it a try.

msfjarvis replied:
```
> Task :app:minifyReleaseWithR8
R8: Missing class: java.lang.ClassValue
```

R8's already having a fit working through Guava's size. Would certainly love not needing it.

FabianHenneke replied:
I have completed the switch over to Mozilla's Public Suffix List and removed the Guava dependency.

I chose to reference the relevant Maven artifact with a `+` version. Although this should usually be avoided, it seemed reasonable to me for a library that is basically set in stone apart from a large list that is updated biweekly. If you don't like to have an unspecified version, feel free to fix one though, I did not give it much thought from the maintainer's perspective.

msfjarvis commented:
I've resolved the trivial issues I pointed out in my review to let you work on the more important parts. Will let you know as soon as I have a mental model for the onboarding UI. Also need to decide what I want to do about our old Autofill...

FabianHenneke commented:
I have resolved the remaining comment and also added a few more changes that should make testing easier.

> I've resolved the trivial issues I pointed out in my review to let you work on the more important parts. Will let you know as soon as I have a mental model for the onboarding UI. Also need to decide what I want to do about our old Autofill...

A random thought of mine regarding the onboarding UI: It would be very nice to tell the user which of their installed browsers is compatible with Autofill by cross-checking with the list of trusted browsers. But of course, this is just of the "nice to have" category.

The old Autofill could possible be renamed "Accessibility Fill", "Pre-Oreo Autofill" or "Legacy Autofill".

msfjarvis commented:
> A random thought of mine regarding the onboarding UI: It would be very nice to tell the user which of their installed browsers is compatible with Autofill by cross-checking with the list of trusted browsers. But of course, this is just of the "nice to have" category.

That's a pretty great idea actually. Would love to build this out.

> The old Autofill could possible be renamed "Accessibility Fill", "Pre-Oreo Autofill" or "Legacy Autofill".

Honestly I wanted to just not expose it to users above API 27 and make everybody use the proper Autofill implementation instead.

msfjarvis commented:
Got tired of dancing around Chrome hoping to bend myself into the position that makes Autofill work for more than two minutes and went to Firefox Preview. Absolutely blown away. The browser itself still feels like garbage but Autofill works like a dream, enough to make me switch permanently. Definitely going to cry foul about Chrome in the onboarding UI.

FabianHenneke commented:
> That's a pretty great idea actually. Would love to build this out.

I am currently working on providing this information for the UI.

> Honestly I wanted to just not expose it to users above and on API 27 and make everybody use the proper Autofill implementation instead.

I wasn't sure whether you would want to go this far, but now that you have proposed it: This is the best solution imo.

FabianHenneke commented:
I added two commits that improve the API in FeatureAndTrustDetection.kt (it now has a single public method that cannot be misused) and added a draft of an onboarding UI that displays the support levels of all installed compatible and the default browser (which may not support Autofill).

Please feel free to flesh this out and add browsers you find to be flaky.

IvanMalison commented:
> Got tired of dancing around Chrome hoping to bend myself into the position that makes Autofill work for more than two minutes and went to Firefox Preview. Absolutely blown away. The browser itself still feels like garbage but Autofill works like a dream, enough to make me switch permanently. Definitely going to cry foul about Chrome in the onboarding UI.

Wow it's that bad? How did google fuck this up that badly?

https://bugs.chromium.org/p/chromium/issues/detail?id=1015381

Does mention (comment 19 https://bugs.chromium.org/p/chromium/issues/detail?id=1015381#c19) that it is working with the chrome beta.

Should we ping the issue?

msfjarvis commented:
> I added two commits that improve the API in FeatureAndTrustDetection.kt (it now has a single public method that cannot be misused) and added a draft of an onboarding UI that displays the support levels of all installed compatible and the default browser (which may not support Autofill).

Sweet!

> Please feel free to flesh this out and add browsers you find to be flaky.
> 
> Edit: Please also check that this doesn't break if a user previously had the old Autofill enabled. We might want to show special UI in that case, currently they would have to disable and re-enable the Autofill setting.

Yes that's going to be an important transition. I was thinking of having a check that shows a dialog when the app is launched for the first time post update and lets people know that the old Autofill is gone and they can opt into this much better version.


> > Got tired of dancing around Chrome hoping to bend myself into the position that makes Autofill work for more than two minutes and went to Firefox Preview. Absolutely blown away. The browser itself still feels like garbage but Autofill works like a dream, enough to make me switch permanently. Definitely going to cry foul about Chrome in the onboarding UI.
> 
> Wow it's that bad? How did google fuck this up that badly?

I have absolutely no idea. I can't get Autofill to work in any way with Chrome Stable, on any of my devices from Android 9 to R DP2.

> [bugs.chromium.org/p/chromium/issues/detail?id=1015381](https://bugs.chromium.org/p/chromium/issues/detail?id=1015381)
> 
> Does mention (comment 19 [bugs.chromium.org/p/chromium/issues/detail?id=1015381#c19](https://bugs.chromium.org/p/chromium/issues/detail?id=1015381#c19)) that it is working with the chrome beta.
> 
> Should we ping the issue?

The last comment is on March 7 so I'm not gonna add to the noise. I've starred the issue.

FabianHenneke commented:
> The last comment is on March 7 so I'm not gonna add to the noise. I've starred the issue.

I think that the relevant commit that mostly fixes the behavior in Chrome Beta is this one: https://chromium-review.googlesource.com/c/chromium/src/+/1980739. Maybe we should wait out the next stable release before we advertise the Autofill feature more widely. That way, we wouldn't have to warn about the flakiness at all.

msfjarvis commented:
> > The last comment is on March 7 so I'm not gonna add to the noise. I've starred the issue.
> 
> I think that the relevant commit that mostly fixes the behavior in Chrome Beta is this one: https://chromium-review.googlesource.com/c/chromium/src/+/1980739. Maybe we should wait out the next stable release before we advertise the Autofill feature more widely. That way, we wouldn't have to warn about the flakiness at all.

Yeah waiting seems like a wise choice to me.

FabianHenneke commented:
I had to fix a bug in the application of field tie breakers and made some improvements to the fillable field detection. As a result, fill and save work well with Google's two-step login now.

In the process of working on these commits, two potentially contentious questions came up:

1. When Password Store detects a submitted login form and offers to save the credentials, should the username be used as the suggested filename? This does allow saving the entry with a single tap, but may increase the likelihood that a username that should only be stored encrypted is accidentally stored unencrypted. Since the latter should only really be an issue for a very small of fraction of users, I have added the feature, but still like to know what you think about it.

2. accounts.google.com (and other websites) use two-step login flows in which the first screen contains a visible username and a hidden password field. If Password Store were to fill the hidden password field, it would come prefilled on the second screen and would save the user the hassle of invoking Autofill again. Assuming that this behavior would be limited to browsers that correctly report the origin on all fields and situations where there is a single password field (and it is hidden), would you think that users would be okay with filling into hidden password fields?

IvanMalison commented:
> I think that the relevant commit that mostly fixes the behavior in Chrome Beta is this one: https://chromium-review.googlesource.com/c/chromium/src/+/1980739. Maybe we should wait out the next stable release before we advertise the Autofill feature more widely. That way, we wouldn't have to warn about the flakiness at all.

Okay, I trust your decision making here, but for the record, I, as an end user, would strongly prefer an earlier release. This feature feels kinda

msfjarvis commented:
> > I think that the relevant commit that mostly fixes the behavior in Chrome Beta is this one: [chromium-review.googlesource.com/c/chromium/src/+/1980739](https://chromium-review.googlesource.com/c/chromium/src/+/1980739). Maybe we should wait out the next stable release before we advertise the Autofill feature more widely. That way, we wouldn't have to warn about the flakiness at all.
> 
> Okay, I trust your decision making here, but for the record, I, as an end user, would strongly prefer an earlier release. This feature feels kinda

I totally understand your desire for this feature to ship at the earliest but the flakiness in question here is more "absolutely unusable" than flaky. Chrome has enough market share for this to escalate into a nightmare on the Play Store review section from any users who don't pay attention to the flakiness warning. If Google delays Chrome releases due to the COVID-19 outbreak which makes waiting untenable then I'm more than happy to reconsider my decision, but for now this is going to stay on hold.

msfjarvis commented:
> I had to fix a bug in the application of field tie breakers and made some improvements to the fillable field detection. As a result, fill and save work well with Google's two-step login now.
> 
> In the process of working on these commits, two potentially contentious questions came up:
> 
> 1. When Password Store detects a submitted login form and offers to save the credentials, should the username be used as the suggested filename? This does allow saving the entry with a single tap, but may increase the likelihood that a username that should only be stored encrypted is accidentally stored unencrypted. Since the latter should only really be an issue for a very small of fraction of users, I have added the feature, but still like to know what you think about it.

I've very recently taken to organising my store to follow the filename-as-username format and found it to be incredibly helpful. We should retain this as-is.

> 2. accounts.google.com (and other websites) use two-step login flows in which the first screen contains a visible username and a hidden password field. If Password Store were to fill the hidden password field, it would come prefilled on the second screen and would save the user the hassle of invoking Autofill again. Assuming that this behavior would be limited to browsers that correctly report the origin on all fields and situations where there is a single password field (and it is hidden), would you think that users would be okay with filling into hidden password fields?

Speaking as an incredibly lazy person, I'd love for the hidden field to be autofilled but I can also see why it could be scary for many. Wouldn't really recommend making this the default behavior. Fine by me to have it gated behind a preference.

FabianHenneke commented:
> I've very recently taken to organising my store to follow the filename-as-username format and found it to be incredibly helpful. We should retain this as-is.

I have added a switch to the save dialog that allows the user to toggle between encrypting the username (i.e. storing it in the extras) and not encrypting it (i.e. using it as the filename; this is the default). I think that this is what should suit most people's usecase, as opposed to a general SharedPref: Most usernames are simply their email address and therefore not very sensitive, but some (e.g. bank account numbers used as logins) are sensitive.

> Speaking as an incredibly lazy person, I'd love for the hidden field to be autofilled but I can also see why it could be scary for many. Wouldn't really recommend making this the default behavior. Fine by me to have it gated behind a preference.

I have overhauled the way hidden fields are treated. The DSL is now able to reason about them, which makes it much more practical to add rules for fine-grained hidden field filling/saving. As a result, I have now chosen to fill into a hidden password field assuming it 1) directly follows a username field with autocomplete hint, 2) does not have autocomplete=off set and 3) no other scenario matches.

msfjarvis commented:
> > I've very recently taken to organising my store to follow the filename-as-username format and found it to be incredibly helpful. We should retain this as-is.
> 
> I have added a switch to the save dialog that allows the user to toggle between encrypting the username (i.e. storing it in the extras) and not encrypting it (i.e. using it as the filename; this is the default). I think that this is what should suit most people's usecase, as opposed to a general SharedPref: Most usernames are simply their email address and therefore not very sensitive, but some (e.g. bank account numbers used as logins) are sensitive.

That sounds perfect :+1: 

> > Speaking as an incredibly lazy person, I'd love for the hidden field to be autofilled but I can also see why it could be scary for many. Wouldn't really recommend making this the default behavior. Fine by me to have it gated behind a preference.
> 
> I have overhauled the way hidden fields are treated. The DSL is now able to reason about them, which makes it much more practical to add rules for fine-grained hidden field filling/saving. As a result, I have now chosen to fill into a hidden password field assuming it 1) directly follows a username field with autocomplete hint, 2) does not have autocomplete=off set and 3) no other scenario matches.


I think the reasoning is accurate and properly balanced between security and convenience.

I'm going to do one final review pass right now. If you're willing to maintain this feature for the foreseeable future then please update the [CODEOWNERS](.github/CODEOWNERS) file as per the [syntax definition](https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners#codeowners-syntax), otherwise let me know so I can merge this and get some real world testing going.

FabianHenneke commented:
> I'm going to do one final review pass right now. If you're willing to maintain this feature for the foreseeable future then please update the [CODEOWNERS](.github/CODEOWNERS) file as per the [syntax definition](https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners#codeowners-syntax), otherwise let me know so I can merge this and get some real world testing going.

I will add myself as a code owner for the relevant directory and will happily maintain the Autofill feature. In that case, would you want me to do merge? 

In any case, I would like to ask you (and the others in this thread) to assist me with testing.

msfjarvis commented:
> > I'm going to do one final review pass right now. If you're willing to maintain this feature for the foreseeable future then please update the [CODEOWNERS](.github/CODEOWNERS) file as per the [syntax definition](https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners#codeowners-syntax), otherwise let me know so I can merge this and get some real world testing going.
> 
> I will add myself as a code owner for the relevant directory and will happily maintain the Autofill feature. In that case, would you want me to do merge?

Yeah, I'll grant you push access.

> In any case, I would like to ask you (and the others in this thread) to assist me with testing.

I've been keeping myself in sync with this PR and testing on all my devices. Our next release is sufficiently far off for us to be able to test this off master so let's get this in.

FabianHenneke commented on app/src/main/java/com/zeapo/pwdstore/autofill/oreo/AutofillHelper.kt:
```diff
@@ -0,0 +1,178 @@
+ */
+val AssistStructure.ViewNode.webOrigin: String?
+    @RequiresApi(Build.VERSION_CODES.O) get() = webDomain?.let { domain ->
+        val scheme = (if (Build.VERSION.SDK_INT >= 28) webScheme else null) ?: "http"
```
I think its not, preventing the user from interacting with login forms on insecure origins is clearly the browser's task. I will go ahead and change this.

FabianHenneke commented on app/src/main/java/com/zeapo/pwdstore/UserPreference.kt:
```diff
@@ -480,16 +533,27 @@ class UserPreference : AppCompatActivity() {
-    private val isServiceEnabled: Boolean
+    private val isAccessibilityServiceEnabled: Boolean
         get() {
             val am = this
```
This was in the old code, now that I have removed it, I also switched to the typed getSystemService overload.

FabianHenneke commented:
I have added myself as a code owner for the oreo/ folder and a wildcard for `*oreo_autofill*`. I renamed the single file that didn't follow that naming scheme.

I also moved up the "Enable Autofill" preference in the settings to give it more visibility at least before the onboarding UI is finalized. Are you going to take on the notification about the new feature?

I will now resolve the conflict and merge.

FabianHenneke commented:
I have resolved the merge conflict and will not make any further changes for now, but would need a new approving review.

FabianHenneke commented:
@msfjarvis Based on my tests, Autofill is still flaky with Chrome Beta from time to time. Should we add it to the list of flaky apps?

msfjarvis commented:
> @msfjarvis Based on my tests, Autofill is still flaky with Chrome Beta from time to time. Should we add it to the list of flaky apps?

I've been having a rather positive experience in the last week or so (which I am sure I just jinxed) but yeah let's add Chrome to the flaky list for now and revisit at a later date when the world is back to normal (whatever that is) and Chromium releases resume.

IvanMalison commented:
@msfjarvis When will this make it i to a release?

msfjarvis commented:
> @msfjarvis When will this make it i to a release?

There's a v1.7.0 milestone set on this PR which has it's closing date, that's the approximate time we aim for that month's release.

> And if it wont be soon, is there any way you could provide an apk of the current master so people who want to try it out can do so without having to download android studio and compile it themselves?

The README has a link for snapshot builds.

IvanMalison commented:
> The README has a link for snapshot builds.

Link seems to be dead.

msfjarvis commented:
> > The README has a link for snapshot builds.
> 
> Link seems to be dead.

Yeah I'm stupid and overlooked the URL format change after my switch to a different file indexing software. Here's the working link: https://dl.msfjarvis.dev/?dir=APS

jvanbruegge commented:
I just tried the snapshot and the search does not work apparently for the naming scheme `path/to/domain.tld.gpg`. For example when trying to log into paypal, the search correctly contains `paypal.com` but it does not find `Banking/paypal.com.gpg`. Clearing the search and selecting the entry from the list however correctly fills username and password and if i save the association I get it suggested the next time above the option to search

msfjarvis commented:
> I just tried the snapshot and the search does not work apparently for the naming scheme `path/to/domain.tld.gpg`. For example when trying to log into paypal, the search correctly contains `paypal.com` but it does not find `Banking/paypal.com.gpg`. Clearing the search and selecting the entry from the list however correctly fills username and password and if i save the association I get it suggested the next time above the option to search

Can you create a separate issue for this? That'll make it easier to triage. Thanks!
